### ISSUE TITLE
Question: Sizing field

### ISSUE THREAD

**Issue by krober10nd** [nschloe/pygalmesh#68]
Hey, Is it possible to pass a gridded interpolant  describing the variation of sizing in the meshing domain? The below example does not work and states:

TypeError: metaclass conflict: the metaclass of a derived class must be a (non-strict) subclass of the metaclasses of all its bases
```python
import pygalmesh

# bottom left and top right corner coords
domain = pygalmesh.Cuboid([bbox[0], 0.0, 0.0], [0.0, bbox[3], bbox[5]])

# define a cell_size function based on an interpolant (scipy.regulargriddedInterpolant)
class Field(pygalmesh.SizingFieldBase, fh):
    def eval(self, x):
        return fh(x)


eff = Field(pygalmesh.SizingFieldBase, fh)

# call the generator
mesh = pygalmesh.generate_with_sizing_field(domain, cell_size=eff,)
import meshio

meshio.write("out.vtk", mesh)
print("finished!")
```
Thanks for any help if possible

**Comment by nschloe** [comment]
Please provide a complete, running MWE.

**Comment by krober10nd** [comment]
Thank you for your response. Unfortunately it's not possible to provide a minimal working example, which is why I'm kindly posing the question. 

Can the user create a pygal.SizingFieldBase that accepts mesh sizing field information from a gridded interpolant? 

In the example on the readme, it is shown how to build a SizingField with analytical functions using point coordinates. However, often in geophysical applications the variation in mesh sizes cannot be described by simple functions like these instead the user creates more complicated sizing fields a stores them as interpoalnts and would like to pass these to the generator. 

This is a more complete example, but as I said earlier it does not work because of the usage of the gridded interpolant.  Note I realize the sizing function is trivial here but it's just to illustrate the functionality that I would like to know more about. 

Thanks

 from scipy.interpolate import RegularGridInterpolator
 import numpy as np
 import pygalmesh
 
 bbox = (-4200, 0, 0, 13520, 0, 13520)
 nx=676
 ny=676
 nz=210
 
 # create a mesh size function interpolant
 zvec = np.linspace(bbox[0], 0, nz, dtype=np.float32)
 xvec = np.linspace(0, bbox[3], nx, dtype=np.float32)
 yvec = np.linspace(0, bbox[5], ny, dtype=np.float32)

 hmin = 50
 hh_m = np.zeros(shape=(nz, nx, ny), dtype=np.float32) + hmin
 interpolant = RegularGridInterpolator(
     (z_vec, x_vec, y_vec), hh_m, bounds_error=False, fill_value=None
 )
 fh = lambda p: interpolant(p)
 
 c1 = pygalmesh.Cuboid([0., 0., 0.], [bbox[0], bbox[3], bbox[5]])
 
 # Here I would like to define a cell_size function based on the gridded interpolant (scipy.regularGriddedInterpolant). However, this doesn't work per the error message in the OP. 
 class Field(pygalmesh.SizingFieldBase, fh):
     def eval(self, x):
         return fh(x)
 
 eff = Field(pygalmesh.SizingFieldBase, fh)
 
 mesh = pygalmesh.generate_with_sizing_field(domain, cell_size=eff,)
 
 import meshio
 meshio.write("out.vtk", mesh)

**Comment by nschloe** [comment]
```
NameError: name 'z_vec' is not defined
```
I want to see the error you're seeing. Right now I see an undefined variable which I'm sure you're not talking about.

**Comment by krober10nd** [comment]
Ah sorry, I updated the code above z_vec should have been zvec..

**Comment by nschloe** [comment]
As a side-note, in the future please make sure that your code is copy-pastable. (It has an incorrect leading white space right now.)

**Comment by nschloe** [comment]
I'm now getting
```
Traceback (most recent call last):
  File "g.py", line 25, in <module>
    class Field(pygalmesh.SizingFieldBase, fh):
TypeError: metaclass conflict: the metaclass of a derived class must be a (non-strict) subclass of the metaclasses of all its bases
```
Again, an unrelated error. :-1:

**Comment by krober10nd** [comment]
Ah no. That's the error I started the thread with.

**Comment by nschloe** [comment]
Ah, well, that's just a "syntax" error in the class.
```python
from scipy.interpolate import RegularGridInterpolator
import numpy as np
import pygalmesh

bbox = (-4200, 0, 0, 13520, 0, 13520)
nx = 676
ny = 676
nz = 210

# create a mesh size function interpolant
zvec = np.linspace(bbox[0], 0, nz, dtype=np.float32)
xvec = np.linspace(0, bbox[3], nx, dtype=np.float32)
yvec = np.linspace(0, bbox[5], ny, dtype=np.float32)

hmin = 50
hh_m = np.zeros(shape=(nz, nx, ny), dtype=np.float32) + hmin
interpolant = RegularGridInterpolator(
    (zvec, xvec, yvec), hh_m, bounds_error=False, fill_value=None
)
fh = lambda p: interpolant(p)

domain = pygalmesh.Cuboid([0.0, 0.0, 0.0], [bbox[0], bbox[3], bbox[5]])

# Here I would like to define a cell_size function based on the gridded interpolant (scipy.regularGriddedInterpolant). However, this doesn't work per the error message in the OP.
class Field(pygalmesh.SizingFieldBase):
    def eval(self, x):
        return fh(x)


eff = Field()

mesh = pygalmesh.generate_with_sizing_field(domain, cell_size=eff,)

import meshio

meshio.write("out.vtk", mesh)
```

**Comment by krober10nd** [comment]
What does that mean? Which class? What's throwing the incorrect syntax error?

**Comment by nschloe** [comment]
I posted the correct code.

**Comment by krober10nd** [comment]
I really appreciate your help, however that doesn't work. I get an error (below). Regardless, without the call to the gridded interpolant, returning a constant value in the sizing Field produces the same error. 

Lets take a step back, what's the arguments for the domain i.e.:

domain = pygalmesh.Cuboid([0.0, 0.0, 0.0], [bbox[0], bbox[3], bbox[5]])

Is the first argument the lower left hand corner and the second the top right (i.e., two points to a diagonal?) I just guessed in this example, but perhaps this is an underlying problem here.

Thanks!


Inserting protection balls...
  refine_balls = true
  min_balls_radius = 0
  min_balls_weight = 0
CGAL error: assertion violation!
Expression : 
File       : /usr/local/include/CGAL/Mesh_3/Protect_edges_sizing_field.h
Line       : 448
Explanation: Error: the sizing field is null at corner (-4200 0 0)
Refer to the bug-reporting instructions at https://www.cgal.org/bug_report.html
Traceback (most recent call last):
  File "test_pygal.py", line 29, in <module>
    mesh = pygalmesh.generate_with_sizing_field(domain, cell_size=Field(), )
  File "/Applications/miniconda3/lib/python3.7/site-packages/pygalmesh-0.6.0-py3.7-macosx-10.9-x86_64.egg/pygalmesh/main.py", line 96, in generate_with_sizing_field
    verbose=verbose,
RuntimeError: Caught an unknown exception!

**Comment by nschloe** [comment]
Try running without a variable sizing field to see you get the domain you want.
The error
```
Explanation: Error: the sizing field is null at corner (-4200 0 0)
```
sounds like there's something wrong with your field.

**Comment by krober10nd** [comment]
I tested a constant field as I mentioned in my previous post. It didn't work. 

I asked what are the inputs supposed to be for the Domain.cuboid constructor? I cannot find documentation on this so I'm guessing that the inputs to the cuboid constructor are point coordinates of opposing corners?

For example, this doesn't work for a unit cube with constant resolution... 

```python
import pygalmesh
import meshio

class Field(pygalmesh.SizingFieldBase):
     def eval(self, x):
         return 0.1
 
mesh = pygalmesh.generate_with_sizing_field( pygalmesh.Cuboid( [0,0,0],[1,1,1]),cell_size=Field() )
 
meshio.write("out.vtk", mesh)
```

Here's the error: 
```
Inserting protection balls...
  refine_balls = true
  min_balls_radius = 0
  min_balls_weight = 0
CGAL error: assertion violation!
Expression : 
File       : /usr/local/include/CGAL/Mesh_3/Protect_edges_sizing_field.h
Line       : 448
Explanation: Error: the sizing field is null at corner (0 0 0)
Refer to the bug-reporting instructions at https://www.cgal.org/bug_report.html
Traceback (most recent call last):
  File "junk.py", line 8, in <module>
    mesh = pygalmesh.generate_with_sizing_field( pygalmesh.Cuboid( [0,0,0],[1,1,1]),cell_size=Field() )
  File "/Applications/miniconda3/lib/python3.7/site-packages/pygalmesh-0.6.0-py3.7-macosx-10.9-x86_64.egg/pygalmesh/main.py", line 96, in generate_with_sizing_field
    verbose=verbose,
RuntimeError: CGAL ERROR: assertion violation!
File: /usr/local/include/CGAL/Mesh_3/Protect_edges_sizing_field.h
Line: 448
Explanation: Error: the sizing field is null at corner (0 0 0)
```

**Comment by nschloe** [comment]
No `generate_with_sizing_field`, just `generate`!

**Comment by krober10nd** [comment]
?. I'm simplifying the variable sizing function problem to see what the problem actually is. If I can't generate the mesh with a constant field how can I with a variable field?  I'm using the exact format you supplied on the readme.

What are the inputs to the cuboid domain?

**Comment by nschloe** [comment]
Use `generate` to make sure you got the `Cuboid` part right.

**Comment by krober10nd** [comment]
...What are correct inputs are for the cuboid domain constructor? Inputting the corner coordinates of the cuboid do not work for generate_mesh or generate_with_sizing_field.

**Comment by nschloe** [comment]
https://github.com/nschloe/pygalmesh/blob/798f800512d915f4f1ea8d596a0a4fe131af5981/src/primitives.hpp#L53

**Comment by nschloe** [comment]
https://github.com/nschloe/pygalmesh/blob/5f0291b929701572aff2c677ce447a2668dfbb2b/test/test_volume_mesh.py#L136

**Comment by krober10nd** [comment]
Yes, I read your code but there's no clear explanation as to what these coordinates are. 
Maybe add a line to describe what the input arguments are in plain english? 

 Anyway, I got it work through a sequence of trial and error. 

 The cuboid domain turned out to be fine but I had to add an ``edge_size`` argument to the call to ```generate_with_sizing_field``` otherwise it doesn't work.  ```cell_size``` alone doesn't work with the cuboid domain.
